runtime.c

```diff
@@ -519,8 +521,44 @@ Class class_setSuperclass(Class cls, Class newSuper)
 	CHECK_ARG(cls);
 	CHECK_ARG(newSuper);
 	if (Nil == cls) { return Nil; }
+
+	LOCK_RUNTIME_FOR_SCOPE();
+
+	safe_remove_from_subclass_list(cls);
+
 	Class oldSuper = cls->super_class;
 	cls->super_class = newSuper;
+
+	// The super class's subclass list is used in certain method resolution scenarios.
+	cls->sibling_class = cls->super_class->subclass_list;
+	cls->super_class->subclass_list = cls;
+
+	if (!class_isMetaClass(cls))
+	{
+		// Update the metaclass's superclass.
+		class_setSuperclass(cls->isa, newSuper->isa);
+	}
+	else
+	{
+		// newSuper is presumably a metaclass. Its isa will therefore be the appropriate root metaclass.
+		cls->isa = newSuper->isa;
+	}
+
+	// Make sure the superclass is initialized if we're initialized.
+	if (objc_test_class_flag(cls, objc_class_flag_initialized))
+	{
+		objc_send_initialize(newSuper);
+		// Update the class's dtable to reflect its new superclass's dtable.
+		if (cls->dtable != uninstalled_dtable)
+		{
+			// we can't use objc_update_dtable_for_class here, as it doesn't take into account
+			// superclasses. It only walks downward.
+			free_dtable(cls->dtable);
```

Comment:
This doesn't look safe.  If this call races with a call to `objc_msgSend` on this object / class in another thread then we're deleting the dtable out from under the other thread.  I think that we need to use something that updates the dtable in place to avoid that.  The entries in the dtable should be safe to (relaxed consistency) atomically replace.